File: app-mobile/src/api/payments.ts
/** Payments API: Stripe Connect onboarding + dues payment intents — routers/payments.py.
 *
 * Card data never touches our servers (SPEC §8.7) — this module only fetches the
 * params the Stripe SDK needs; the PaymentSheet flow lives in src/payments/dues.tsx.
 */

import { request } from "./client";

/**
 * Which rail the member is paying on. Chosen BEFORE the intent exists because
 * Stripe freezes application_fee_amount at creation, and the platform fee differs
 * per rail (1% card, 2% ACH).
 */
export type PaymentRail = "card" | "ach";

/** Response of POST /payments/connect/onboarding-link. */
export interface OnboardingLinkOut {
  url: string;
  expires_at: string;
}

/** Response of GET /chapters/{chapter_id}/payments/status. */
export interface ChapterPaymentsStatus {
  onboarded: boolean;
  charges_enabled: boolean;
  details_submitted: boolean;
}

/**
 * Where the intent behind this response already stands with Stripe (board c234).
 *
 * "awaiting_payment" is the default and is what a freshly created intent always
 * reports: the client secret is safe to open in PaymentSheet. The other two mean
 * a same-rail retry retrieved an intent Stripe has already moved past waiting on
 * the member (an ACH debit in flight, or money already taken) while our webhook
 * has not resolved the reservation yet. The client secret is still real, but
 * presenting it as a fresh checkout would ask someone to pay a second time.
 */
export type DuesIntentStatus = "awaiting_payment" | "processing" | "succeeded";

/** Response of POST /payments/dues/{cycle_id}/intent — everything PaymentSheet needs. */
export interface DuesIntentOut {
  payment_intent_client_secret: string;
  customer_session_client_secret: string;
  customer_id: string;
  publishable_key: string;
  stripe_account_id: string;
  amount_cents: number;
  application_fee_cents: number;
  rail: PaymentRail;
  payment_intent_status: DuesIntentStatus;
}

/** Platform cut per rail in basis points, mirroring stripe_service.PLATFORM_FEE_BPS. */
const PLATFORM_FEE_BPS: Record<PaymentRail, number> = { card: 100, ach: 200 };

export function platformFeeCents(amountCents: number, rail: PaymentRail): number {
  return Math.floor((amountCents * PLATFORM_FEE_BPS[rail]) / 10_000);
}

export async function createOnboardingLink(chapterId: string): Promise<OnboardingLinkOut> {
  return request<OnboardingLinkOut>("/payments/connect/onboarding-link", {
    method: "POST",
    body: { chapter_id: chapterId },
  });
}

export async function getChapterPaymentsStatus(
  chapterId: string,
): Promise<ChapterPaymentsStatus> {
  return request<ChapterPaymentsStatus>(`/chapters/${chapterId}/payments/status`);
}

export async function createDuesPaymentIntent(
  cycleId: string,
  rail: PaymentRail,
  amountCents = 0,
): Promise<DuesIntentOut> {
  return request<DuesIntentOut>(`/payments/dues/${cycleId}/intent`, {
    method: "POST",
    body: { rail },
  });
}

File: backend/app/schemas/payments.py
"""Payment schemas: Connect onboarding, account status, dues PaymentSheet params."""
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

# Which rail the member chose. application_fee_amount is immutable once a
# PaymentIntent exists, so the rail is picked BEFORE the intent is created —
# otherwise there is no moment where the fee and the rail are both known.
PaymentRail = Literal["card", "ach"]


class ConnectOnboardingRequest(BaseModel):
    """Body for POST /payments/connect/onboarding-link."""

    chapter_id: uuid.UUID


class ConnectOnboardingOut(BaseModel):
    """Stripe-hosted Express onboarding link; single-use and short-lived."""

    url: str
    expires_at: datetime


class ChapterPaymentsStatusOut(BaseModel):
    """Whether a chapter can actually accept dues yet."""

    model_config = ConfigDict(from_attributes=True)

    onboarded: bool
    charges_enabled: bool
    details_submitted: bool


class DuesIntentCreate(BaseModel):
    """Body for POST /payments/dues/{cycle_id}/intent."""

    rail: PaymentRail


class DuesIntentOut(BaseModel):
    """Everything the mobile PaymentSheet needs for one dues payment.

    amount_cents is echoed from the dues cycle (server-side) so the client can
    render the total it is about to charge — it is never an input.
    """

    payment_intent_client_secret: str
    customer_session_client_secret: str
    customer_id: str
    publishable_key: str
    stripe_account_id: str
    amount_cents: int
    application_fee_cents: int
    rail: PaymentRail
    # Additive (board c234): a same-rail retry retrieves whatever intent the
    # reservation already points at, which can have moved past "awaiting payment"
    # while the client was away (Stripe settled it, or an ACH debit is mid-flight).
    # 'awaiting_payment' — the default, and always what a freshly created intent
    # reports — means payment_intent_client_secret above is safe to open in
    # PaymentSheet. 'processing'/'succeeded' mean it is NOT: the client must show
    # an honest "already completed/processing" message instead.
    payment_intent_status: Literal["awaiting_payment", "processing", "succeeded"] = (
        "awaiting_payment"
    )
